pkg/executor/util.go

```diff
@@ -127,6 +134,10 @@ func labelsForService(
 			result[name] = value
 		}
 	}
+	// Apply all cluster labels also
+	for name, val := range cr.Labels {
+		result[name] = val
+	}
```

Comment:
I'm not truly opposed to this but it does raise some questions. E.g.
* Why apply these to just services (and not pods, pvcs, statefulsets, whatever)?
* Does this raise the expectation that these labels would track the kdcluster's labels (whereas we're actually just doing a one-time copy)?

Was there a specific usecase that this helps? (And something that couldn't be satisfied by first determining the kdcluster and then going to look at its labels.)

Reply:
Yes, there is a specific use case in MLOps tenant since we branch clusters into Notebook, Deployment, and Training and we would like to segregate service Endpoints accordingly. That is where services are different than other resources like Pod, statefulset, etc since services are directly consumed by Enduser and user needs to know more about them, the same is not true for Pods, stateful sets, deployments, etc. If need be we can continue to propagate on other resources as well.

Reply:
Apparently the current UI design makes it hard to lookup by determining the kdcluster and then going to look at its labels. @Nabin and @Rajus can comment more on this

Reply:
Got it. OK, let me summarize my understanding of the current label-related changes here.

* Whenever any statefulset is created by us, we copy to it all current labels on the kdapp. This also includes setting the labels in the statefulset pod templates, which the statefulset will later use to create pods.

* Whenever any service is created by us, we copy to it all current labels on the kdapp and all current labels on the kdcluster.

This feels a little shaky to me. For example let's say that a kdapp has some label foo=bar and we create a kdcluster from it that has 1 role with 1 member. The kdcluster has the label name=joel.

We'll end up with:
```
headless service labels foo=bar, name=joel
statefulset 0    labels foo=bar, name=joel
pod 0-0          labels foo=bar, name=joel
service 0-0      labels foo=bar, name=joel
```
Now we edit the kdapp label to be foo=zowie, and then we expand the role to 2 members.
```
headless service labels foo=bar, name=joel
statefulset 0    labels foo=bar, name=joel
pod 0-0          labels foo=bar, name=joel
service 0-0      labels foo=bar, name=joel
pod 0-1          labels foo=bar, name=joel
service 0-1      labels foo=zowie, name=joel
```
Now we edit the kdcluster label to be name=kartik, and we add a member in another role.
```
headless service labels foo=bar, name=joel
statefulset 0    labels foo=bar, name=joel
pod 0-0          labels foo=bar, name=joel
service 0-0      labels foo=bar, name=joel
pod 0-1          labels foo=bar, name=joel
service 0-1      labels foo=zowie, name=joel
statefulset 1    labels foo=zowie, name=kartik
pod 1-0          labels foo=zowie, name=kartik
service 1-0      labels foo=zowie, name=kartik
```
If at this point we edited the kdapp label back to foo=bar and expanded the second role, we'd have elements of this one kdcluster with all four possible combinations of those label values.

I think some label propagation from the kdapp to the kdcluster and its components would be good. But we have to be careful about how & when we snapshot them, unless we're going to commit to tracking them for reconciliation.

I don't think that it's sufficient to assume that the labels will never be edited. And we can't choose to enforce that either, unless we limit ourselves to a very particular label key.

Hmm.

==========

To be sure I understand the usecase before proceeding:

We have labels on the kdapps, and we want to get those labels onto the services.

Is there no need to get those labels onto the kdcluster resources too?

Is it just a very specific "app category" label that we care about?

Reply:
hmm, actually the current requirement is just to get the kdcluster labels to service. That should be ok. Currently, the UI is applying some labels on kdcluster which I thought could potentially be avoided by propagating all the way from app to cluster to services. If I look at above example then you are right it can get a bit hairy. So it's getting unnecessarily complicated for now so  I will leave it for now

Reply:
The label we care about right now is createdByUser which UI is adding on kdcluster. Based on that the UI will display only the Notebook cluster and service endpoints for the logged-in user.

Reply:
The UI could explicitly put that in the serviceLabels in the role config of the kdcluster...

Reply:
hmmm, that should be ok too, so you are suggesting lets table the label propagation for now and revisit later since labels can change?

Reply:
yup, if possible

Reply:
ok, I will work with UI folks. We will come back to this if we hit a stalemate.